src/value.jl

```diff
@@ -1,145 +1,167 @@
-function CategoricalValue(level::Integer, pool::CategoricalPool{T, R}) where {T, R}
-    return CategoricalValue(convert(R, level), pool)
-end
-
-Base.convert(::Type{CategoricalValue{T, R}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{CategoricalValue{T}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{CategoricalValue}, x::CategoricalValue) = x
-
-Base.convert(::Type{Union{CategoricalValue{T, R}, Null}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{Union{CategoricalValue{T}, Null}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
-
-Base.promote_rule(::Type{CategoricalValue{S, R}}, ::Type{T}) where {S, T, R} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue{S}}, ::Type{T}) where {S, T} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue}, ::Type{T}) where {T} = T
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# checks whether the type is "categorical value"
+iscatvalue(::Type) = false
+iscatvalue(::Type{Union{}}) = false # otherwise it dispatches to Type{<:...}
+iscatvalue(::Type{<:CatValue}) = true
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+valtype(::Type{<:CategoricalValue{T}}) where {T} = T
+valtype(::Type{<:CategoricalString}) = String
+
+# integer type of category reference codes used by categorical value
+reftype(::Type{<:CatValue{R}}) where {R} = R
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvalue_type(::Type{T}) where {T} = T
+unwrap_catvalue_type(::Type{T}) where {T >: Null} =
+    Union{unwrap_catvalue_type(Nulls.T(T)), Null}
+unwrap_catvalue_type(::Type{Any}) = Any # to prevent dispatching to T>:Null method
```

Comment:
Is that really a problem? `Union{Any, Null} == Any`. Same question for the `catvalue_type` method below.